File: src/lib/projectNotes.ts
import { invoke } from "@tauri-apps/api/core";

import { isHostedWebBrowserMode } from "./mockOrchestra/host";
import { getHostedWebOrchestraClientBinding } from "./orchestraClient/runtime";
import { getProject } from "./projects";
import type { NoteDetail, NoteLocation, NoteTreeNode, NotesRoot, NotesTree, ProjectDetail } from "../types";

const MOCK_NOTES_STORAGE_KEY = "orchestra.mock.notes";

type MockScopeNotes = {
  notes: Record<string, string>;
  directories: string[];
};

type MockNotesState = Record<string, MockScopeNotes>;

type DirectoryAccumulator = {
  name: string;
  path: string;
  directories: Map<string, DirectoryAccumulator>;
  notes: Map<string, NoteTreeNode>;
};

function isTauriAvailable() {
  return Boolean((window as Window & { __TAURI_INTERNALS__?: unknown }).__TAURI_INTERNALS__);
}

function getHostedWebClient() {
  return getHostedWebOrchestraClientBinding()?.client ?? null;
}

function loadMockNotesState(): MockNotesState {
  const raw = window.localStorage.getItem(MOCK_NOTES_STORAGE_KEY);
  return raw ? JSON.parse(raw) as MockNotesState : {};
}

function saveMockNotesState(state: MockNotesState) {
  window.localStorage.setItem(MOCK_NOTES_STORAGE_KEY, JSON.stringify(state));
}

function normalizePath(path: string, options?: { allowEmpty?: boolean; requireMarkdown?: boolean }) {
  const allowEmpty = options?.allowEmpty ?? false;
  const requireMarkdown = options?.requireMarkdown ?? false;
  const raw = path.trim().replace(/\\/g, "/");
  if (!raw && allowEmpty) {
    return "";
  }
  if (!raw) {
    throw new Error("path: Path is required.");
  }
  if (raw.startsWith("/") || raw.startsWith("~/")) {
    throw new Error("path: Note paths must stay relative to docs/.");
  }

  const parts = raw.split("/").filter(Boolean);
  for (const part of parts) {
    if (part === "." || part === "..") {
      throw new Error("path: Note paths must stay inside docs/.");
    }
  }

  const normalized = parts.join("/");
  if (!normalized && !allowEmpty) {
    throw new Error("path: Path is required.");
  }
  if (requireMarkdown && normalized && !normalized.toLowerCase().endsWith(".md")) {
    throw new Error("path: Note files must end with .md.");
  }
  return normalized;
}

function normalizeLocation(location: NoteLocation, options?: { allowEmpty?: boolean; requireMarkdown?: boolean }) {
  const scope = location.scope === "repository" ? "repository" : "project";
  const repositoryId = scope === "repository" ? location.repositoryId?.trim() ?? "" : "";
  if (scope === "repository" && !repositoryId) {
    throw new Error("repositoryId: Repository note locations require repositoryId.");
  }
  return {
    scope,
    repositoryId: repositoryId || null,
    path: normalizePath(location.path, options),
  } satisfies NoteLocation;
}

function scopeStorageKey(projectId: string, location: NoteLocation) {
  return location.scope === "project"
    ? `${projectId}::project`
    : `${projectId}::repository:${location.repositoryId}`;
}

function ensureScopeState(state: MockNotesState, key: string): MockScopeNotes {
  if (!state[key]) {
    state[key] = { notes: {}, directories: [] };
  }
  return state[key]!;
}

function addParentDirectories(entry: MockScopeNotes, path: string) {
  const segments = path.split("/").filter(Boolean);
  const next = new Set(entry.directories);
  for (let index = 1; index < segments.length; index += 1) {
    next.add(segments.slice(0, index).join("/"));
  }
  entry.directories = Array.from(next).sort();
}

function deleteDirectoryContents(entry: MockScopeNotes, path: string) {
  const prefix = path ? `${path}/` : "";
  entry.directories = entry.directories.filter((directory) => directory !== path && !directory.startsWith(prefix));
  for (const notePath of Object.keys(entry.notes)) {
    if (notePath === path || notePath.startsWith(prefix)) {
      delete entry.notes[notePath];
    }
  }
}

function copyDirectoryContents(source: MockScopeNotes, sourcePath: string, destination: MockScopeNotes, destinationPath: string) {
  const sourcePrefix = sourcePath ? `${sourcePath}/` : "";
  const destinationPrefix = destinationPath ? `${destinationPath}/` : "";
  const nextDirectories = new Set(destination.directories);

  if (destinationPath) {
    nextDirectories.add(destinationPath);
    addParentDirectories(destination, destinationPath);
  }

  for (const directory of source.directories) {
    if (directory === sourcePath || directory.startsWith(sourcePrefix)) {
      const suffix = directory === sourcePath ? "" : directory.slice(sourcePrefix.length);
      const nextDirectory = [destinationPath, suffix].filter(Boolean).join("/");
      if (nextDirectory) {
        nextDirectories.add(nextDirectory);
      }
    }
  }

  for (const [notePath, markdown] of Object.entries(source.notes)) {
    if (notePath === sourcePath || notePath.startsWith(sourcePrefix)) {
      const suffix = notePath === sourcePath ? "" : notePath.slice(sourcePrefix.length);
      const nextPath = [destinationPath, suffix].filter(Boolean).join("/");
      destination.notes[nextPath] = markdown;
      addParentDirectories(destination, nextPath);
    }
  }

  destination.directories = Array.from(nextDirectories).sort();
}

function createDirectoryRoot(): DirectoryAccumulator {
  return {
    name: "",
    path: "",
    directories: new Map(),
    notes: new Map(),
  };
}

function ensureDirectory(root: DirectoryAccumulator, directoryPath: string) {
  let current = root;
  const segments = directoryPath.split("/").filter(Boolean);
  const built: string[] = [];
  for (const segment of segments) {
    built.push(segment);
    const currentPath = built.join("/");
    const existing = current.directories.get(segment);
    if (existing) {
      current = existing;
      continue;
    }
    const next: DirectoryAccumulator = {
      name: segment,
      path: currentPath,
      directories: new Map(),
      notes: new Map(),
    };
    current.directories.set(segment, next);
    current = next;
  }
  return current;
}

function buildTreeNodes(entry: MockScopeNotes): NoteTreeNode[] {
  const root = createDirectoryRoot();
  for (const directory of entry.directories) {
    if (directory) {
      ensureDirectory(root, directory);
    }
  }

  for (const notePath of Object.keys(entry.notes)) {
    const segments = notePath.split("/").filter(Boolean);
    const fileName = segments.pop();
    if (!fileName) {
      continue;
    }
    const parent = segments.length ? ensureDirectory(root, segments.join("/")) : root;
    parent.notes.set(fileName, {
      kind: "note",
      name: fileName,
      path: notePath,
    });
  }

  const toNodes = (directory: DirectoryAccumulator): NoteTreeNode[] => {
    const directoryNodes = Array.from(directory.directories.values())
      .sort((left, right) => left.name.localeCompare(right.name))
      .map((child) => ({
        kind: "directory" as const,
        name: child.name,
        path: child.path,
        children: toNodes(child),
      }));
    const noteNodes = Array.from(directory.notes.values())
      .sort((left, right) => left.name.localeCompare(right.name));
    return [...directoryNodes, ...noteNodes];
  };

  return toNodes(root);
}

async function loadProjectDetail(projectId: string): Promise<ProjectDetail> {
  const project = await getProject(projectId);
  if (!project) {
    throw new Error(`Project ${projectId} was not found.`);
  }
  return project;
}

function getRootState(projectId: string, location: NoteLocation, state: MockNotesState) {
  const key = scopeStorageKey(projectId, location);
  return ensureScopeState(state, key);
}

function assertLocationBelongsToProject(project: ProjectDetail, location: NoteLocation) {
  if (location.scope === "project") {
    return;
  }
  const repositoryId = location.repositoryId ?? "";
  if (!project.repositories.some((repository) => repository.id === repositoryId)) {
    throw new Error(`Repository ${repositoryId} does not belong to project ${project.id}.`);
  }
}

export async function listProjectNotes(projectId: string): Promise<NotesTree> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.list(projectId);
  }
  if (isTauriAvailable()) {
    return invoke<NotesTree>("list_project_notes", { projectId });
  }

  const project = await loadProjectDetail(projectId);
  const state = loadMockNotesState();
  const projectLocation: NoteLocation = { scope: "project", path: "" };
  const projectEntry = getRootState(projectId, projectLocation, state);
  const repositoryRoots = [...project.repositories]
    .sort((left, right) => left.name.localeCompare(right.name))
    .map((repository) => {
      const location: NoteLocation = { scope: "repository", repositoryId: repository.id, path: "" };
      const entry = getRootState(projectId, location, state);
      return {
        scope: "repository" as const,
        repositoryId: repository.id,
        label: repository.name,
        docsExists: Boolean(Object.keys(entry.notes).length || entry.directories.length),
        children: buildTreeNodes(entry),
      } satisfies NotesRoot;
    });

  return {
    projectId,
    roots: [
      {
        scope: "project",
        repositoryId: null,
        label: "Project",
        docsExists: Boolean(Object.keys(projectEntry.notes).length || projectEntry.directories.length),
        children: buildTreeNodes(projectEntry),
      },
      ...repositoryRoots,
    ],
  };
}

export async function getProjectNote(projectId: string, location: NoteLocation): Promise<NoteDetail> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.get(projectId, location);
  }
  if (isTauriAvailable()) {
    return invoke<NoteDetail>("get_project_note", { projectId, location });
  }

  const project = await loadProjectDetail(projectId);
  const normalized = normalizeLocation(location, { requireMarkdown: true });
  assertLocationBelongsToProject(project, normalized);
  const state = loadMockNotesState();
  const entry = getRootState(projectId, normalized, state);
  const markdown = entry.notes[normalized.path];
  return {
    location: normalized,
    markdown: markdown ?? "",
    exists: markdown !== undefined,
  };
}

export async function updateProjectNote(projectId: string, location: NoteLocation, markdown: string): Promise<NoteDetail> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.update(projectId, location, markdown);
  }
  if (isTauriAvailable()) {
    return invoke<NoteDetail>("update_project_note", { projectId, location, markdown });
  }

  const project = await loadProjectDetail(projectId);
  const normalized = normalizeLocation(location, { requireMarkdown: true });
  assertLocationBelongsToProject(project, normalized);
  const state = loadMockNotesState();
  const entry = getRootState(projectId, normalized, state);
  entry.notes[normalized.path] = markdown;
  addParentDirectories(entry, normalized.path);
  saveMockNotesState(state);
  return { location: normalized, markdown, exists: true };
}

export async function deleteProjectNote(projectId: string, location: NoteLocation): Promise<NoteLocation> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.delete(projectId, location);
  }
  if (isTauriAvailable()) {
    return invoke<NoteLocation>("delete_project_note", { projectId, location });
  }

  const project = await loadProjectDetail(projectId);
  const normalized = normalizeLocation(location, { requireMarkdown: true });
  assertLocationBelongsToProject(project, normalized);
  const state = loadMockNotesState();
  const entry = getRootState(projectId, normalized, state);
  delete entry.notes[normalized.path];
  saveMockNotesState(state);
  return normalized;
}

export async function copyProjectNote(projectId: string, source: NoteLocation, destination: NoteLocation): Promise<NoteDetail> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.copy(projectId, source, destination);
  }
  if (isTauriAvailable()) {
    return invoke<NoteDetail>("copy_project_note", { projectId, source, destination });
  }

  const project = await loadProjectDetail(projectId);
  const normalizedSource = normalizeLocation(source, { requireMarkdown: true });
  const normalizedDestination = normalizeLocation(destination, { requireMarkdown: true });
  assertLocationBelongsToProject(project, normalizedSource);
  assertLocationBelongsToProject(project, normalizedDestination);
  const state = loadMockNotesState();
  const sourceEntry = getRootState(projectId, normalizedSource, state);
  const markdown = sourceEntry.notes[normalizedSource.path];
  if (markdown === undefined) {
    throw new Error(`Note ${normalizedSource.path} was not found.`);
  }
  const destinationEntry = getRootState(projectId, normalizedDestination, state);
  destinationEntry.notes[normalizedDestination.path] = markdown;
  addParentDirectories(destinationEntry, normalizedDestination.path);
  saveMockNotesState(state);
  return { location: normalizedDestination, markdown, exists: true };
}

export async function moveProjectNote(projectId: string, source: NoteLocation, destination: NoteLocation): Promise<NoteDetail> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.move(projectId, source, destination);
  }
  if (isTauriAvailable()) {
    return invoke<NoteDetail>("move_project_note", { projectId, source, destination });
  }

  const copied = await copyProjectNote(projectId, source, destination);
  await deleteProjectNote(projectId, source);
  return copied;
}

export async function createProjectNotesDirectory(projectId: string, location: NoteLocation): Promise<NoteLocation> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.createDirectory(projectId, location);
  }
  if (isTauriAvailable()) {
    return invoke<NoteLocation>("create_project_notes_directory", { projectId, location });
  }

  const project = await loadProjectDetail(projectId);
  const normalized = normalizeLocation(location, { allowEmpty: true });
  assertLocationBelongsToProject(project, normalized);
  const state = loadMockNotesState();
  const entry = getRootState(projectId, normalized, state);
  const nextDirectories = new Set(entry.directories);
  if (normalized.path) {
    nextDirectories.add(normalized.path);
    addParentDirectories(entry, normalized.path);
  }
  entry.directories = Array.from(nextDirectories).sort();
  saveMockNotesState(state);
  return normalized;
}

export async function deleteProjectNotesDirectory(projectId: string, location: NoteLocation): Promise<NoteLocation> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.deleteDirectory(projectId, location);
  }
  if (isTauriAvailable()) {
    return invoke<NoteLocation>("delete_project_notes_directory", { projectId, location });
  }

  const project = await loadProjectDetail(projectId);
  const normalized = normalizeLocation(location, { allowEmpty: true });
  assertLocationBelongsToProject(project, normalized);
  const state = loadMockNotesState();
  const entry = getRootState(projectId, normalized, state);
  deleteDirectoryContents(entry, normalized.path);
  saveMockNotesState(state);
  return normalized;
}

export async function copyProjectNotesDirectory(projectId: string, source: NoteLocation, destination: NoteLocation): Promise<NoteLocation> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.copyDirectory(projectId, source, destination);
  }
  if (isTauriAvailable()) {
    return invoke<NoteLocation>("copy_project_notes_directory", { projectId, source, destination });
  }

  const project = await loadProjectDetail(projectId);
  const normalizedSource = normalizeLocation(source, { allowEmpty: true });
  const normalizedDestination = normalizeLocation(destination, { allowEmpty: true });
  assertLocationBelongsToProject(project, normalizedSource);
  assertLocationBelongsToProject(project, normalizedDestination);
  if (
    normalizedSource.scope === normalizedDestination.scope
    && normalizedSource.repositoryId === normalizedDestination.repositoryId
    && normalizedDestination.path
    && (normalizedDestination.path === normalizedSource.path || normalizedDestination.path.startsWith(`${normalizedSource.path}/`))
  ) {
    throw new Error("Destination directory cannot be the same directory or a descendant of the source directory.");
  }
  const state = loadMockNotesState();
  const sourceEntry = getRootState(projectId, normalizedSource, state);
  const destinationEntry = getRootState(projectId, normalizedDestination, state);
  copyDirectoryContents(sourceEntry, normalizedSource.path, destinationEntry, normalizedDestination.path);
  saveMockNotesState(state);
  return normalizedDestination;
}

export async function moveProjectNotesDirectory(projectId: string, source: NoteLocation, destination: NoteLocation): Promise<NoteLocation> {
  const hostedWebClient = getHostedWebClient();
  if (hostedWebClient) {
    return hostedWebClient.notes.moveDirectory(projectId, source, destination);
  }
  if (isTauriAvailable()) {
    return invoke<NoteLocation>("move_project_notes_directory", { projectId, source, destination });
  }

  const moved = await copyProjectNotesDirectory(projectId, source, destination);
  await deleteProjectNotesDirectory(projectId, source);
  return moved;
}

File: src-tauri/src/services/project_notes.rs
use std::{
    fs,
    path::{Path, PathBuf},
};

use rusqlite::Connection;
use uuid::Uuid;

use crate::{
    models::{NoteDetail, NoteLocation, NoteTreeNode, NotesRoot, NotesTree},
    services::{
        orchestra_paths::{default_orchestra_root, managed_repository_checkout_dir, project_root},
        projects,
    },
};

pub fn list_notes(connection: &Connection, project_id: &str) -> Result<NotesTree, String> {
    list_project_notes(connection, project_id)
}

pub fn get_note(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
) -> Result<NoteDetail, String> {
    get_project_note(connection, project_id, location.clone())
}

pub fn update_note(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
    markdown: &str,
) -> Result<NoteDetail, String> {
    update_project_note(
        connection,
        project_id,
        location.clone(),
        markdown.to_string(),
    )
}

pub fn delete_note(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
) -> Result<NoteLocation, String> {
    delete_project_note(connection, project_id, location.clone())
}

pub fn copy_note(
    connection: &Connection,
    project_id: &str,
    source: &NoteLocation,
    destination: &NoteLocation,
) -> Result<NoteDetail, String> {
    copy_project_note(connection, project_id, source.clone(), destination.clone())
}

pub fn move_note(
    connection: &Connection,
    project_id: &str,
    source: &NoteLocation,
    destination: &NoteLocation,
) -> Result<NoteDetail, String> {
    move_project_note(connection, project_id, source.clone(), destination.clone())
}

pub fn create_directory(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
) -> Result<NoteLocation, String> {
    create_project_notes_directory(connection, project_id, location.clone())
}

pub fn delete_directory(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
) -> Result<NoteLocation, String> {
    delete_project_notes_directory(connection, project_id, location.clone())
}

pub fn copy_directory(
    connection: &Connection,
    project_id: &str,
    source: &NoteLocation,
    destination: &NoteLocation,
) -> Result<NoteLocation, String> {
    copy_project_notes_directory(connection, project_id, source.clone(), destination.clone())
}

pub fn move_directory(
    connection: &Connection,
    project_id: &str,
    source: &NoteLocation,
    destination: &NoteLocation,
) -> Result<NoteLocation, String> {
    move_project_notes_directory(connection, project_id, source.clone(), destination.clone())
}

pub fn list_project_notes(connection: &Connection, project_id: &str) -> Result<NotesTree, String> {
    let project = projects::get_project(connection, project_id)?;
    let orchestra_root = default_orchestra_root()?;
    let mut roots = vec![build_notes_root(
        "Project",
        NoteLocation {
            scope: "project".into(),
            repository_id: None,
            path: String::new(),
        },
        &resolve_scope_root(
            connection,
            project_id,
            &NoteLocation {
                scope: "project".into(),
                repository_id: None,
                path: String::new(),
            },
        )?,
    )?];

    let mut repositories = project.repositories;
    repositories.sort_by(|left, right| left.name.to_lowercase().cmp(&right.name.to_lowercase()));
    for repository in repositories {
        let checkout_root = repository
            .repository_path
            .clone()
            .map(PathBuf::from)
            .unwrap_or_else(|| {
                managed_repository_checkout_dir(&orchestra_root, &project.slug, &repository.slug)
            });
        roots.push(build_notes_root(
            &repository.name,
            NoteLocation {
                scope: "repository".into(),
                repository_id: Some(repository.id.clone()),
                path: String::new(),
            },
            &checkout_root.join("docs"),
        )?);
    }

    Ok(NotesTree {
        project_id: project.id,
        roots,
    })
}

pub fn get_project_note(
    connection: &Connection,
    project_id: &str,
    location: NoteLocation,
) -> Result<NoteDetail, String> {
    let normalized = normalize_note_location(&location, true, true)?;
    let path = resolve_location_path(connection, project_id, &normalized)?;
    match fs::read_to_string(&path) {
        Ok(markdown) => Ok(NoteDetail {
            location: normalized,
            markdown,
            exists: true,
        }),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(NoteDetail {
            location: normalized,
            markdown: String::new(),
            exists: false,
        }),
        Err(error) => Err(format!("Unable to read note {}: {error}", path.display())),
    }
}

pub fn update_project_note(
    connection: &Connection,
    project_id: &str,
    location: NoteLocation,
    markdown: String,
) -> Result<NoteDetail, String> {
    let normalized = normalize_note_location(&location, true, true)?;
    let path = resolve_location_path(connection, project_id, &normalized)?;
    write_file_atomically(&path, &markdown)?;
    Ok(NoteDetail {
        location: normalized,
        markdown,
        exists: true,
    })
}

pub fn delete_project_note(
    connection: &Connection,
    project_id: &str,
    location: NoteLocation,
) -> Result<NoteLocation, String> {
    let normalized = normalize_note_location(&location, true, true)?;
    let path = resolve_location_path(connection, project_id, &normalized)?;
    match fs::remove_file(&path) {
        Ok(()) => Ok(normalized),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(normalized),
        Err(error) => Err(format!("Unable to delete note {}: {error}", path.display())),
    }
}

pub fn copy_project_note(
    connection: &Connection,
    project_id: &str,
    source: NoteLocation,
    destination: NoteLocation,
) -> Result<NoteDetail, String> {
    let normalized_source = normalize_note_location(&source, true, true)?;
    let normalized_destination = normalize_note_location(&destination, true, true)?;
    let source_path = resolve_location_path(connection, project_id, &normalized_source)?;
    let destination_path = resolve_location_path(connection, project_id, &normalized_destination)?;
    let markdown = fs::read_to_string(&source_path).map_err(|error| {
        format!(
            "Unable to read source note {}: {error}",
            source_path.display()
        )
    })?;
    write_file_atomically(&destination_path, &markdown)?;
    Ok(NoteDetail {
        location: normalized_destination,
        markdown,
        exists: true,
    })
}

pub fn move_project_note(
    connection: &Connection,
    project_id: &str,
    source: NoteLocation,
    destination: NoteLocation,
) -> Result<NoteDetail, String> {
    let normalized_source = normalize_note_location(&source, true, true)?;
    let normalized_destination = normalize_note_location(&destination, true, true)?;
    let copied = copy_project_note(
        connection,
        project_id,
        normalized_source.clone(),
        normalized_destination.clone(),
    )?;
    delete_project_note(connection, project_id, normalized_source)?;
    Ok(copied)
}

pub fn create_project_notes_directory(
    connection: &Connection,
    project_id: &str,
    location: NoteLocation,
) -> Result<NoteLocation, String> {
    let normalized = normalize_note_location(&location, true, false)?;
    if normalized.path.is_empty() {
        return Ok(normalized);
    }
    let path = resolve_location_path(connection, project_id, &normalized)?;
    fs::create_dir_all(&path).map_err(|error| {
        format!(
            "Unable to create note directory {}: {error}",
            path.display()
        )
    })?;
    Ok(normalized)
}

pub fn delete_project_notes_directory(
    connection: &Connection,
    project_id: &str,
    location: NoteLocation,
) -> Result<NoteLocation, String> {
    let normalized = normalize_note_location(&location, false, false)?;
    let path = resolve_location_path(connection, project_id, &normalized)?;
    match fs::remove_dir_all(&path) {
        Ok(()) => Ok(normalized),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(normalized),
        Err(error) => Err(format!(
            "Unable to delete note directory {}: {error}",
            path.display()
        )),
    }
}

pub fn copy_project_notes_directory(
    connection: &Connection,
    project_id: &str,
    source: NoteLocation,
    destination: NoteLocation,
) -> Result<NoteLocation, String> {
    let normalized_source = normalize_note_location(&source, false, false)?;
    let normalized_destination = normalize_note_location(&destination, false, false)?;
    validate_directory_move(&normalized_source, &normalized_destination)?;
    let source_path = resolve_location_path(connection, project_id, &normalized_source)?;
    let destination_path = resolve_location_path(connection, project_id, &normalized_destination)?;
    copy_directory_recursively(&source_path, &destination_path)?;
    Ok(normalized_destination)
}

pub fn move_project_notes_directory(
    connection: &Connection,
    project_id: &str,
    source: NoteLocation,
    destination: NoteLocation,
) -> Result<NoteLocation, String> {
    let normalized_source = normalize_note_location(&source, false, false)?;
    let normalized_destination = normalize_note_location(&destination, false, false)?;
    validate_directory_move(&normalized_source, &normalized_destination)?;
    copy_project_notes_directory(
        connection,
        project_id,
        normalized_source.clone(),
        normalized_destination.clone(),
    )?;
    delete_project_notes_directory(connection, project_id, normalized_source)?;
    Ok(normalized_destination)
}

fn build_notes_root(
    label: &str,
    location: NoteLocation,
    docs_root: &Path,
) -> Result<NotesRoot, String> {
    Ok(NotesRoot {
        scope: location.scope,
        repository_id: location.repository_id,
        label: label.to_string(),
        docs_exists: docs_root.exists(),
        children: list_tree_nodes(docs_root, "")?,
    })
}

fn list_tree_nodes(root: &Path, relative_path: &str) -> Result<Vec<NoteTreeNode>, String> {
    if !root.exists() {
        return Ok(Vec::new());
    }
    let mut directories = Vec::new();
    let mut notes = Vec::new();
    for entry in fs::read_dir(root)
        .map_err(|error| format!("Unable to read notes directory {}: {error}", root.display()))?
    {
        let entry = entry.map_err(|error| format!("Unable to read notes entry: {error}"))?;
        let file_type = entry.file_type().map_err(|error| {
            format!(
                "Unable to inspect notes entry {}: {error}",
                entry.path().display()
            )
        })?;
        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') {
            continue;
        }
        let path = if relative_path.is_empty() {
            name.clone()
        } else {
            format!("{relative_path}/{name}")
        };
        if file_type.is_dir() {
            directories.push(NoteTreeNode {
                kind: "directory".into(),
                name: name.clone(),
                path: path.clone(),
                children: Some(list_tree_nodes(&entry.path(), &path)?),
            });
        } else if file_type.is_file() && name.to_lowercase().ends_with(".md") {
            notes.push(NoteTreeNode {
                kind: "note".into(),
                name,
                path,
                children: None,
            });
        }
    }
    directories.sort_by(|left, right| left.name.to_lowercase().cmp(&right.name.to_lowercase()));
    notes.sort_by(|left, right| left.name.to_lowercase().cmp(&right.name.to_lowercase()));
    directories.extend(notes);
    Ok(directories)
}

fn resolve_scope_root(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
) -> Result<PathBuf, String> {
    let project = projects::get_project(connection, project_id)?;
    let orchestra_root = default_orchestra_root()?;
    if location.scope == "project" {
        return Ok(project_root(&orchestra_root, &project.slug).join("docs"));
    }

    let repository_id = location
        .repository_id
        .as_deref()
        .ok_or_else(|| "repositoryId: Repository notes require repositoryId.".to_string())?;
    let repository = project
        .repositories
        .iter()
        .find(|entry| entry.id == repository_id)
        .ok_or_else(|| {
            format!("Repository {repository_id} does not belong to project {project_id}")
        })?;
    let checkout_root = repository
        .repository_path
        .clone()
        .map(PathBuf::from)
        .unwrap_or_else(|| {
            managed_repository_checkout_dir(&orchestra_root, &project.slug, &repository.slug)
        });
    Ok(checkout_root.join("docs"))
}

fn resolve_location_path(
    connection: &Connection,
    project_id: &str,
    location: &NoteLocation,
) -> Result<PathBuf, String> {
    let root = resolve_scope_root(connection, project_id, location)?;
    if location.path.is_empty() {
        return Ok(root);
    }
    Ok(root.join(location.path.split('/').collect::<PathBuf>()))
}

fn normalize_note_location(
    location: &NoteLocation,
    allow_empty: bool,
    require_markdown: bool,
) -> Result<NoteLocation, String> {
    let scope = match location.scope.trim() {
        "project" => "project",
        "repository" => "repository",
        other => return Err(format!("scope: Unsupported note scope {other}.")),
    };
    let repository_id = location
        .repository_id
        .as_ref()
        .map(|value| value.trim())
        .filter(|value| !value.is_empty())
        .map(str::to_string);
    if scope == "repository" && repository_id.is_none() {
        return Err("repositoryId: Repository notes require repositoryId.".into());
    }
    let path = normalize_relative_docs_path(&location.path, allow_empty, require_markdown)?;
    Ok(NoteLocation {
        scope: scope.into(),
        repository_id: repository_id,
        path,
    })
}

fn normalize_relative_docs_path(
    value: &str,
    allow_empty: bool,
    require_markdown: bool,
) -> Result<String, String> {
    let trimmed = value.trim().replace('\\', "/");
    if trimmed.is_empty() {
        if allow_empty {
            return Ok(String::new());
        }
        return Err("path: Path is required.".into());
    }
    if trimmed.starts_with('/') || trimmed.starts_with("~/") {
        return Err("path: Paths must stay relative to docs/.".into());
    }
    let mut parts = Vec::new();
    for part in trimmed.split('/') {
        let part = part.trim();
        if part.is_empty() || part == "." {
            continue;
        }
        if part == ".." {
            return Err("path: Paths must stay inside docs/.".into());
        }
        parts.push(part);
    }
    if parts.is_empty() && !allow_empty {
        return Err("path: Path is required.".into());
    }
    let normalized = parts.join("/");
    if require_markdown && !normalized.to_lowercase().ends_with(".md") {
        return Err("path: Note paths must end with .md.".into());
    }
    Ok(normalized)
}

fn validate_directory_move(
    source: &NoteLocation,
    destination: &NoteLocation,
) -> Result<(), String> {
    if source.path.is_empty() {
        return Err("path: Root docs directories cannot be copied or moved as directories.".into());
    }
    if source.scope == destination.scope
        && source.repository_id == destination.repository_id
        && (destination.path == source.path
            || destination.path.starts_with(&format!("{}/", source.path)))
    {
        return Err("path: Destination cannot be the same directory or a descendant of the source directory.".into());
    }
    Ok(())
}

fn write_file_atomically(path: &Path, markdown: &str) -> Result<(), String> {
    let parent = path
        .parent()
        .ok_or_else(|| format!("Path {} has no parent directory", path.display()))?;
    fs::create_dir_all(parent).map_err(|error| {
        format!(
            "Unable to create note directory {}: {error}",
            parent.display()
        )
    })?;
    let temp_path = parent.join(format!(
        ".{}.tmp-{}-{}",
        path.file_name()
            .and_then(|value| value.to_str())
            .unwrap_or("note.md"),
        std::process::id(),
        Uuid::new_v4().simple()
    ));
    fs::write(&temp_path, markdown).map_err(|error| {
        format!(
            "Unable to write note temp file {}: {error}",
            temp_path.display()
        )
    })?;
    if path.exists() {
        fs::remove_file(path)
            .map_err(|error| format!("Unable to replace note {}: {error}", path.display()))?;
    }
    fs::rename(&temp_path, path).map_err(|error| {
        format!(
            "Unable to move note temp file {} into place at {}: {error}",
            temp_path.display(),
            path.display()
        )
    })?;
    Ok(())
}

fn copy_directory_recursively(source: &Path, destination: &Path) -> Result<(), String> {
    if !source.exists() {
        return Err(format!(
            "Source directory {} was not found.",
            source.display()
        ));
    }
    fs::create_dir_all(destination).map_err(|error| {
        format!(
            "Unable to create destination directory {}: {error}",
            destination.display()
        )
    })?;
    for entry in fs::read_dir(source).map_err(|error| {
        format!(
            "Unable to read source directory {}: {error}",
            source.display()
        )
    })? {
        let entry = entry.map_err(|error| format!("Unable to read directory entry: {error}"))?;
        let file_type = entry.file_type().map_err(|error| {
            format!(
                "Unable to inspect directory entry {}: {error}",
                entry.path().display()
            )
        })?;
        let entry_name = entry.file_name();
        let source_path = entry.path();
        let destination_path = destination.join(entry_name);
        if file_type.is_dir() {
            copy_directory_recursively(&source_path, &destination_path)?;
        } else if file_type.is_file() {
            if let Some(parent) = destination_path.parent() {
                fs::create_dir_all(parent).map_err(|error| {
                    format!(
                        "Unable to create destination directory {}: {error}",
                        parent.display()
                    )
                })?;
            }
            fs::copy(&source_path, &destination_path).map_err(|error| {
                format!(
                    "Unable to copy {} to {}: {error}",
                    source_path.display(),
                    destination_path.display()
                )
            })?;
        }
    }
    Ok(())
}
